### ISSUE TITLE
what's the /src/locales folder?

### ISSUE THREAD

**Issue by birbilis** [concord-consortium/lab#123]
I thought this repo was for the engine, however I see it contains folder with some localized models

Shouldn't those go to 
https://github.com/birbilis/lab-interactives-site/tree/master/src/locales/es/interactives
instead?

Does the engine itself have some resources needing localization and if so, where are they?

**Comment by sfentress** [comment]
Hi @birbilis, the Lab framework has a few demo models, to make it easier to see what the framework does without needing to deal with the main model repository in `lab-interactives-site`. These can be seen in `src/models`, and are visible at the demo site http://lab-framework.concord.org/

The /src/locales folder is for translations of these models, as you guessed. The engine itself also has strings that are translatable (for the tooltips, etc.), and these can be seen in `src/lab/locales/`.

**Comment by birbilis** [comment]
guess you mean https://github.com/birbilis/lab/blob/master/src/lab/locales/translations.json, hadn't noticed that one

how come a single file is used there instead of separate .json files combined with a metadata .json file, like the website does? Would be more manageable when there are many more languages in the future and easier to merge pull requests without conflicts (plus less stuff to load if done on demand and lab is remotely served)

**Comment by birbilis** [comment]
btw, another benefit of keeping as separate files the translation resources is that people could diff the resource file for the language they're interested in maintaining with the English one and see what strings have been added since the first translation to add those too. As it is now in a single file that is very hard to do

**Comment by sfentress** [comment]
Those are fair points. I think our reasoning -- to the extent that it was thought about deeply -- was that the strings file for the Lab framework is tiny in comparison to the translations for the interactives. It's about 230 English strings for the framework, vs hundreds of pages of translatable strings for all the interactives.

However, if this affects your work and you would like to make the changes and submit a Pull Request for it, I'd be happy to look it over.

**Comment by birbilis** [comment]
any hints where the relevant codebits of the lab and the lab-interactives are for me to compare?

**Comment by birbilis** [comment]
btw, at both
"asparticacid": "Asparticacid",
and
"glutamicacid": "Glutamicacid",
shouldn't there be a space before "acid"?

**Comment by sfentress** [comment]
Hi @birbilis, it's been years since I touched any part of this code that relates to translations, and I don't think it would be very efficient for me to try and track down all the places in the code where such a change would touch.

If this feature is interesting to you, I think it might be an good introduction to the code to try and find out all the parts that would need to be changed. I'd be happy to look at specific questions related to a PR in progress.

**Comment by birbilis** [comment]
If I find some time to look at it, I will give it a shot

In the mean time, I added a Greek translation, where I separated language entries with a blank line (guess JSON has no issue with blank lines) so that a translator can browse them more easily when using a simple text editor like the one in GitHub (instead of using a tree-like editor). 
https://github.com/concord-consortium/lab/pull/124

I also added the fix for the two acid names in English that I mention above, since I see other translations do have them as two words.

Another possible issue I see is that there is some text that says something like "Please __click_here_link__" instead of saying "Please click __click_here_link__" as I'd expect (is there a missing "click" word there?), which also confuses on who wants to translate it, since they don't know if they should add the click word in their language. I didn't fix that one, wasn't sure what the __click_here_link__ is (just some URL, or something that contains the word click). If it does contain that English word, how is one supposed to translate that?

**Comment by birbilis** [comment]
an issue I'm having with trying to translate models from lab interactives repo, is that the educators used
https://lab.concord.org/interactives.html#interactives/samples/5-amino-acids.json
page and the "Interactive Editor" pane at the bottom to translate the 5-amino-acids interactive to Greek, but when I search for that one at the github repo for lab interactives, the English version I find on the site is not the same as the one in the repo (there seem to be some missing stuff from the JSON [probably some fields in the repo are left to get default values at runtime?] and others have different order [probably fields of JSON aren't sorted alphabetically at saving to ensure a consistent field order for comparing files]). Should I instruct them to not directly translate interactives online (helps to be able to test translation immediately there), but translate json files only originating from the github repo?

btw at
http://lab.concord.org/doc/models/md2d/index.html
the link "MD2D" pointing to https://github.com/concord-consortium/lab/tree/master/src/lab/md2d
is broken if you can pass the word (not sure if it is an issue with the page or with the repo and what the correct url should be there)

**Comment by ddamelin** [comment]
Hi @birbilis, 
It is fine to translate using the online page. I would just copy and paste the entire thing when creating the translated version of that file.

**Comment by birbilis** [comment]
@ddamelin thanks, so I used the translation of the online version, even though it has some differences from the one in the lab-interactives repo, but getting blocked by Travis. Any idea why?
https://github.com/concord-consortium/lab-interactives-site/pull/67

**Comment by birbilis** [comment]
btw, just above that I see the following text, how come it writes "-ake:" there instead of "-make:" ?

-ake: Entering directory `/home/travis/build/concord-consortium/lab-interactives-site/node_modules/node-inspector/node_modules/socket.io/node_modules/socket.io-client/node_modules/ws/build'

  CXX(target) Release/obj.target/bufferutil/src/bufferutil.o

  SOLINK_MODULE(target) Release/obj.target/bufferutil.node

  SOLINK_MODULE(target) Release/obj.target/bufferutil.node: Finished

  COPY Release/bufferutil.node

  CXX(target) Release/obj.target/validation/src/validation.o

  SOLINK_MODULE(target) Release/obj.target/validation.node

  SOLINK_MODULE(target) Release/obj.target/validation.node: Finished

  COPY Release/validation.node

make: Leaving directory `/home/travis/build/concord-consortium/lab-interactives-site/node_modules/node-inspector/node_modules/socket.io/node_modules/socket.io-client/node_modules/ws/build'

**Comment by ddamelin** [comment]
@birbilis We'll take a look at Travis. Definitely seems like it's a general build problem, not related to your specific pull request.

**Comment by ddamelin** [comment]
@scytacki or @pjanik Can you take a look at what is happening with Travis when it tries to build the lab-interactive-site?

**Comment by pjanik** [comment]
@ddamelin, @birbilis, I'll take a look at it tomorrow morning. It looks like some NPM issue, not related to this PR.

**Comment by birbilis** [comment]
Hi, any news on the travis/npm issue? Will travis retry again for ever or does it timeout and giveup after a number of retries? Do I have to close and reopen the pull request once the configuration issue is fixed for it to try again?

**Comment by pjanik** [comment]
@birbilis, I have a fix ready, just waiting for it to be merged into master. 
When it's done, I'll let you know. The best way would be to update your fork with the latest fixes and rebase your changes. I can help with git commands.

**Comment by pjanik** [comment]
@birbilis, there's a fix in master branch of the lab-interactives repo. You will need to rebase your PR to include this change and then Travis should build your branch.

If you're not sure how to do it:
```
# Add the remote (original lab-interactives-site), call it "upstream":

git remote add upstream https://github.com/concord-consortium/lab-interactives-site.git

# Fetch all the branches of that remote into remote-tracking branches,
# such as upstream/master:

git fetch upstream

# Make sure that you're on your topic branch (I guess it's Greek-translation):

git checkout Greek-translation

# Create a new one, based on this

git checkout Greek-translation-fixed

# Rewrite your master Greek-translation-fixed branch so that any commits of yours that
# aren't already in upstream/master are replayed on top of that
# other branch:

git rebase upstream/master

# Push updated branch

git push origin Greek-translation-fixed
```

and then you can open a new PR.

HTH

**Comment by pjanik** [comment]
@birbilis, I've made two changes:
- I updated country code from `el` to `gr` as that's the official ISO code. `el` didn't have any flag associated with it, so it was looking a bit broken.
- I fixed and merged your PR in the lab-interactives-site. You can see your translation here: http://lab.concord.org/interactives-staging.html#interactives/samples/5-amino-acids.json

If you update your fork to include latest changes from our master, new PR shouldn't have any problems with TravisCI builds.

Is there anything left or can we close this issue?

Thanks!

**Comment by birbilis** [comment]
Actually, el (hellenic) should be the language code and GR (greece) the country specifier. 

Eg 
el-GR = Greek from Greece
el-CY = Greek from Cyprus

Plain "el" should be the correct one I think. Will check again what ISO codes specify and let you know

**Comment by birbilis** [comment]
ISO 639-1 code for Greek is "el" like I had. See https://en.m.wikipedia.org/wiki/Greek_language

Probably mixed it with country code which should be "gr"

Languages aren't necessarily related to countries, dialects and other variations sometimes are (eg Greek Cypriot)

Can you change your system to fix this and revert the change in the pull request?

**Comment by birbilis** [comment]
Btw since translations json says en-US those are definitely language codes, not country codes. It would be bad to use country codes like UK, US anyway. Also I see usage of en-US, where plain en could be used. Unless British English is to be added etc.

Problem also on menus is if you use flags to match a language. Not exactly accurate. For Greek language (el) can use Greek flag anyway, no need to match to el-GR exactly. It isn't usual to see el-CY specific anyway in software or written in official docs, cyprus greek is more of a spoken diversion with some archaic words and pronunciation.

**Comment by birbilis** [comment]
btw, at translation of lab strings I've also used "el" (which as I explain above is the correct language code), so if "gr" is used for a certain localized interactive, then probably lab strings (names of elements etc.) won't show correctly (will show in English). If you don't want to use generic language codes (like "en", "el" etc.) with the flags UI you have on the site (flags aren't exactly appropriate indicators for languages anyway), at least use "el-GR" which is a more specific language code (but have to change it also at the lab strings in this repo and at the one interactive [5-amino-acids] I've submitted to the lab-interactives-site repo)

since this issue can break urls to localized interactives (e.g. #locales/el/interactives/samples/5-amino-acids.json vs #locales/gr/interactives/samples/5-amino-acids.json that you use now), let's resolve it ASAP, since we're in the process of localizing more interactives to Greek and need to be able to send them for review etc. (prefer sending a localized url than telling them to pick the Greek flag from the language selection UI every time they visit the page)

**Comment by birbilis** [comment]
sorry for the long thread

another reason you should use "el" btw is that browsers can give you the list of languages the user has set (in the OS if it's a newer one like Win10, and/or in the browser settings) in preferred order for rendering content. Greeks usually have "el" there, not "el-GR" (which is more specific). Then they usually have "en" (or sometimes "en-US" since they switch between ΕΛ and ENG (USA) keyboard in Windows) as second in the list

see screenshot

![image](https://user-images.githubusercontent.com/3461504/28067206-ea8b6d8e-6648-11e7-9f5a-a2040260adec.png)

then, when user visits an interactive in lab that doesn't have a locale specifier, it could jump to the preferred language of the user if available for that specific interactive. Then, selecting manually the en-US from the flags menu, they could be taken to a #locales/en-US/interactives/... url that has copies (or aliases if webserver allows such) of the interactives in english version (or could instead set some flag that for that interactive's session they want to ignore the preferred language of the browser, so that it doesn't jump back to the preferred language when they pick the en-US [the default])

for the way to detect the visitor's order of preferred languages list, see:
https://www.google.gr/search?q=get+preferred+languages+list+from+browser&ie=utf-8&oe=utf-8&client=firefox-b-ab&gfe_rd=cr&ei=6b1kWZOQGuyAX7-YqpgP

**Comment by pjanik** [comment]
>  INDEED, JUST CHECKED AND THE ELEMENTS SHOW IN ENGLISH

Note that new version of Lab which includes Greek translation is not released yet. So, you need to use staging environment for now:
http://lab.concord.org/interactives-staging.html#interactives/samples/5-amino-acids.json
(-staging segment, environment can be also changed using a pull down in the top section).

I've changed "gr" back to "el" and simply renamed flag from "gr.png" to "el.png", so it doesn't look broken and "el" can be used.

Thanks!

**Comment by birbilis** [comment]
thanks, how long does it take to move from staging to production usually?

**Comment by pjanik** [comment]
I think there's a plan to release a new version very soon, so probably a few days (depends on QA testing).

**Comment by birbilis** [comment]
thanks, I'll send later on today a few changes to Greek strings for the lab then, hope those will be included too

**Comment by pjanik** [comment]
Okay. Updates to lab-interactives-site repo are immediately deploted, but Lab framework is a bit more complicated. I think this thread can be closed now (and new PRs or issues can be opened). Thanks.

**Comment by birbilis** [comment]
@pjanik are your people still on vacations? hoping https://github.com/concord-consortium/lab-interactives-site/pull/70 will be merged soon
